### ISSUE TITLE
Inexact calculations effects on clippings

### ISSUE THREAD

**Issue by skygering** [JuliaGeo/GeometryOps.jl#114]
Here is an example of inexact calculations causing issues:

```julia
scale_fac = 1e5
c1 = [[-28083.868447876892, -58059.13401805979], [-9833.052704767595, -48001.726711609794], [-16111.439295815226, -2.856614689791036e-11], [-76085.95770326033, -2.856614689791036e-11], [-28083.868447876892, -58059.13401805979]]
c1 = [c ./ scale_fac for c in c1]
c2 = [[-53333.333333333336, 0.0], [0.0, 0.0], [0.0, -80000.0], [-60000.0, -80000.0], [-53333.333333333336, 0.0]]
c2 = [c ./ scale_fac for c in c2]
p1 = GI.Polygon([c1])
p2 = GI.Polygon([c2])
GO.intersection(p1, p2; target = GI.PolygonTrait(), fix_multipoly = nothing)
```

When `scale_fac = 1` this doesn't work, but when `scale_fac = 1e5` it does work...

Basically, there are two edges over one another on top of the x-axis, but one of the lines is slightly below the x-axis. I am working on a "patch" for this type of issue, but we might want a stronger fix at some point.

**Comment by asinghvi17** [comment]
We could always add ExactPredicates as a "backend", and use regular floating point computations as the default?  Or the other way around.  It looks like the refactor from earlier should allow for that if I understand correctly?

**Comment by skygering** [comment]
It is the math here though... not comparisons like `ExactPredicates` would be used for. Maybe we can do an exact comparison if the numbers are really close to zero... Or somehow use the [backed package ](https://github.com/JuliaIntervals/IntervalArithmetic.jl)? I don't really understand how `ExactPredicates` works yet...

Also, how do we know when to use it?

What do we think about adding an `atol` parameter that the user can pass in that allows for a bit of fudging to help with this problem for now? This is still problematic and should be fixed long term, but only happens when the polygons are of a large scale and then things are super close to zero... So for example if my polygons above are in meters, I am okay with fudging of up to like 0.5 meters personally for my applications, which would get rid of this problem I think. 

I am drafting a PR with this to give the user more flexibility.

**Comment by asinghvi17** [comment]
Would these issues have happened before #72?  More specifically, is this what you're looking for?
![iTerm2](https://github.com/JuliaGeo/GeometryOps.jl/assets/32143268/a4262b0b-99db-47b7-bc77-a8fea4d9a923)


I got that result without scaling on the commit immediately before #72, so if that's the case then it shows that the exact arithmetic was probably helping here...

**Comment by skygering** [comment]
Yeah that should be it. Ugh! I just don't think we should be calling it all the time. The performance is SO bad on the geometries in #72. 

But this is an example where it would actually help... it is just that Exact Predicates doesn't actually do the math. It just tells us a bit about what type of intersection it is. That doesn't actually help fix everything I don't think... but it would help haha

**Comment by asinghvi17** [comment]
ExactPredicates probably also needs some help becoming faster (though there's only so much type stability you can have in that case :D).  Maybe playing with turning down the threshold for where ExactPredicates falls into adaptive math might help?

In general this is why I was advocating for a `FastMath()`/`ExactMath()` (or maybe even `; exact = _True()`) - though I guess you would need to have it on all the time unless you are 100% sure about how your polygons are conditioned...

**Comment by skygering** [comment]
Yeah... It is hard to know if you are going to need it until you need it. And you only know because the polygons come out weird.

**Comment by rafaqz** [comment]
I think the majority of our exact predicates calls would be removed with judicious extents checks, and sorts or other preparations .

We should try to optimise everything else hard so its less painful to turn it back on.

**Comment by asinghvi17** [comment]
Yeah, preparation should definitely speed this up a lot if LibGEOS is any indicator!

**Comment by skygering** [comment]
Another example:
```
p1 = GI.Polygon([[[-57725.80869813739, -52709.704377648755], [-53333.333333333336, 0.0], [-41878.01362848005, 0.0], [-36022.23699059147, -43787.61366192682], [-48268.44121252392, -52521.18593721105], [-57725.80869813739, -52709.704377648755]]])
p2 = GI.Polygon([[[-60000.0, 80000.0], [0.0, 80000.0], [0.0, 0.0], [-53333.33333333333, 0.0], [-50000.0, 40000.0], [-60000.0, 80000.0]]])
```
<img width="967" alt="Screenshot 2024-04-20 at 9 06 27 AM" src="https://github.com/JuliaGeo/GeometryOps.jl/assets/60117338/edb91647-37dc-4afd-9ef9-d4518e26b7ff">

**Comment by skygering** [comment]
I also did find another bug that maybe this one sets off... but even after fixing that bug this one still doesn't work. So if it errors rather than just breaks, let me know.

**Comment by asinghvi17** [comment]
Are those polygons meant to be unioned or intersected?

**Comment by skygering** [comment]
Intersected. Basically, the problem is that the left-most intersection point is right right next to the orange one's endpoint... by like 1e-11.

**Comment by asinghvi17** [comment]
LibGEOS just returns a linestring for this second scenario - `LINESTRING (-53333.33333333333 0, -41878.01362848005 0)`

**Comment by asinghvi17** [comment]
A more faithful overview of different predicate schemes:
![download-8](https://github.com/JuliaGeo/GeometryOps.jl/assets/32143268/8eafe8ab-60c2-48bd-ae5e-8770e5c669bb)
ExactPredicates is actually >1000x slower, not 50x slower!  Adaptive methods _are_ 10x slower but that itself should speed up polygon operations significantly over ExactPredicates

**Comment by skygering** [comment]
Okay so after our discussion, I think these are the functions we need.

```julia
"""
isparallel((p, q), (a, b))

* Return 1 if the open line segments `(p, q)` and `(a, b)` are parallel.
* Return 0 otherwise.
"""
```

```julia
"""
iscross_nonparallel((p, q), (a, b))

* Return 1 if the non-parallel open line segments `(p, q)` and `(a, b)` fully cross (intersection point isn't an endpoint of either line).
* Return 0 if the non-parallel closed line segments `[p, q]` and `[a, b]` meet at one point in a hinge (the endpoint of at least one of the lines is the intersection point).
* Return –1 otherwise (they don't intersect).
"""
```
```julia
"""
isoverlap_parallel((p, q), (a, b))

* Return 1 if the parallel open line segments `(p, q)` and `(a, b)` overlap (two of the endpoints form a shared region of each line segment).
* Return 0 if the parallel closed line segments `[p, q]` and `[a, b]` meet at one point in a hinge (the endpoint of at least one of the lines is the intersection point).
* Return –1 otherwise (they don't intersect).
"""
```

**Comment by asinghvi17** [comment]
I just saw this comment - have pushed a skeleton of what I think this would look like up to a branch.  Will PR it to yours.

**Comment by skygering** [comment]
Wanted to drop this here: https://github.com/vchuravy/AdaptivePredicates.jl/tree/main

Big thanks to @vchuravy for working on this!

**Comment by asinghvi17** [comment]
Nice!

**Comment by skygering** [comment]
These examples are now working! And performing okay!

This is the top example that I posted in the issue. I will test a little more and then open a PR!
<img width="837" alt="Screenshot 2024-04-24 at 2 19 54 PM" src="https://github.com/JuliaGeo/GeometryOps.jl/assets/60117338/789b2249-ce78-46fe-9313-c870bc34e4f5">

**Comment by asinghvi17** [comment]
Whoa, that's fast!  Guessing LibGEOS is held back by caching/geometry preparation there.

**Comment by skygering** [comment]
```julia
p1 = GI.Polygon([[[-55965.680060140774, -31588.16072168928], [-55956.50771556479, -31478.09258677756], [-31577.548550575284, -6897.015828572996], [-15286.184961223798, -15386.952072224134], [-9074.387601621409, -27468.20712382156], [-8183.4538916097845, -31040.003969070774], [-27011.85123029944, -38229.02388009402], [-54954.72822634951, -32258.9734800704], [-55965.680060140774, -31588.16072168928]]])
p2 = GI.Polygon([[[-80000.0, -80000.0], [-80000.0, 80000.0], [-60000.0, 80000.0], [-50000.0, 40000.0], [-60000.0, -80000.0], [-80000.0, -80000.0]]])
GO.intersection(p1, p2; target = GI.PolygonTrait(), fix_multipoly = nothing)
```
<img width="960" alt="Screenshot 2024-04-26 at 1 16 32 PM" src="https://github.com/JuliaGeo/GeometryOps.jl/assets/60117338/3f1e388c-a1d4-4671-aebe-be2361a73d95">

Another problem... these are touching at basically 1 point, but it is getting registered as two different points since we now have the exact calculations. But they are the "same" number because that is as exact as Floating point can do. 

Although LibGEOS is registering it at a LineString. I am still investigating but I don't want to lose this example so I am commenting right now.

**Comment by skygering** [comment]
I am worried that we should be making this less exact, not more. Maybe we can play with the `atol` values to get approximate answers actually working.

Making some of the calculations more exact has now broken a bunch of the geom relations examples. This is because we have to consider if the midpoint of a line is in or out or on a polygon. The problem is that taking `(a + b) / 2` where `a` and `b` are points, isn't giving an exact midpoint due to floating point... which is now making it not be "on" lines where the exact point actually is due to floating point errors. I don't know how to fix that.

**Comment by asinghvi17** [comment]
So I tried this with Meshes.jl, which is approximate all the way, and it gave an even worse answer:
```
julia> p1 ∩ p2
PolyArea{2,Float64}
  outer
  └─ Ring((-55965.7, -31588.2), (-55965.1, -31581.4), (-55964.6, -31574.7))
```
Maybe we can approximate in some places but keep exact predicates in others?

**Comment by skygering** [comment]
I have no idea how to tell when we should do each of them though... and also it we do approximate then I think we need some `atol` that is related to the size of the geometries. Which feels a little too trial and error for my taste (not to play devil's advocate to myself). But it is concerning that even the most robust of the algorithms (GEOS) struggles with this. It does not inspire confidence in my ability to figure out what is going on just on determination and debugging. 

Also I hate reading C code 😆 spoiled by Julia BUT
- https://github.com/libgeos/geos/blob/main/src/algorithm/Intersection.cpp#L53
- https://github.com/libgeos/geos/blob/main/src/algorithm/Orientation.cpp#L38

There are similar ideas here to what we are doing so that is something at least

**Comment by asinghvi17** [comment]
It looks like LibGEOS is just casting to MultiFloats' `Float64x2` when computing predicates.

**Comment by asinghvi17** [comment]
```julia
julia> GO.touches(p1, p2)
false

julia> GO.touches(p2, p1)
false
```
but 
```julia

julia> p, q = [-50000.0, 40000.0], [-60000.0, -80000.0]
([-50000.0, 40000.0], [-60000.0, -80000.0])

julia> slope = (q[2] - p[2]) / (q[1] - p[1])
12.0

julia> intercept = p[2] - slope * p[1]
640000.0

julia> x = -55965.680060140774
-55965.680060140774

julia> y = slope * x + intercept
-31588.16072168923

julia> y
-31588.16072168923
```
so there's probably an issue there as well

**Comment by skygering** [comment]
Yeah, those returned `true` on the main branch. So that is the issue with making some of the stuff more exact... basically, since it is checking exactness beyond floating point representation, it is really hard to get stuff like touches, where it touches on one point, correct...

**Comment by DanielVandH** [comment]
Did anyone end up getting much further into this issue? Sometimes an annoying issue with dealing with exact predicates is that you do need to carry a few predicate results around sometimes (for example, the length of my `triangle_line_segment_intersection` [here](https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/d3bdb2e9d5518bf00748e26a1d28b745289ed082/src/predicates/predicates.jl#L316-L450) is ridiculous. `line_segment_intersection_type` [here](https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/d3bdb2e9d5518bf00748e26a1d28b745289ed082/src/predicates/predicates.jl#L130-L166) is a bit long too.) I think if you implement it correctly the exactness shouldn't be causing issues as are being described, other than the tedium of sometimes needing to be extra careful with the number of predicates needed (or having to define your own with codegen).

With inexact calculations you are probably always going to be able to find examples where things go very wrong, right?

**Comment by skygering** [comment]
Hey! #146 was merged, which fixes the exact issues within the polygon clipping. However, I still haven't gotten around to extending that to other parts of the code. We rely on `ExactPredicates.jl` right now and are looking to add an intermediate step of adaptive predicates for improved performance. This does take care of the issues described above!

**Comment by asinghvi17** [comment]
Just FYI, I did add AdaptivePredicates in #275 for `orient` which seems like the thing we need most.
